### ISSUE TITLE
Eliminate `rocket_contrib` crate by moving to core under various features

### ISSUE THREAD

**Issue by jhpratt** [SergioBenitez/Rocket#1107]
**TL;DR**: Eliminate the `rocket_contrib` and `rocket_contrib_codegen` crates, merging into `rocket` and `rocket_codegen` respectively. Top-level exports stay the same.

---

### Motivation

Currently, the main `rocket` crate is usable on its own, though `rocket_contrib` is used quite widely. This is not my opinion, `rocket` has 300,963 downloads compared to `rocket_contrib`'s 222,619 at the time of creating this issue, around 74%. In the past 7 days, this averages to 575 per day for `rocket` and 470 per day for `rocket_contrib`, indicating that the current ratio is closer to 82%.

While download numbers alone aren't a great indicator of how commonly the two are used together, I've honestly no idea on how to analyze such a thing on GitHub, for example. As such, my conclusion is that `rocket_contrib` is used in over 4 out of every 5 Rocket projects, and honestly probably should be used in all of them (for helmet if nothing else).

### Proposed changes

Eliminate the `rocket_contrib` crate entirely, deprecating it on crates.io. All of the existing code and features would be moved into the current `core/lib` directory.

I believe that all of the current top-level modules in `rocket_contrib` would be suitable to be top-level modules on `rocket` as well.

The current top-level modules are:

- `json`
- `serve`
- `msgpack`
- `templates`
- `uuid`
- `databases`
- `helmet`

The modules would remain behind feature flags as they currently are, allowing users to opt-out of unnecessarily compilation if they're not using something. The defaults (`json` and `serve`) are sensible, though I'd personally include `helmet` as well.

### What about the `rocket_contrib_codegen`?

Merge it into `rocket_codgen`, eliminating the `rocket_contrib_codegen` crate. It wouldn't make much sense to have a codgen crate for a crate that doesn't exist.

### `rocket_http`?

This could also be merged, though I haven't included it as part of this proposal because it's typically not a direct dependency. I'm not opposed to merging that as well, however.

**Comment by jhpratt** [comment]
Pinging @SergioBenitez and @jebrosen for thoughts. It's a long shot (and likely controversial), but allowing users to only depend on a single crate seems like a noble goal.

**Comment by jebrosen** [comment]
I think it's worth considering moving some of these into `core`, but not yet - IMO certainly not in `0.5`. The default features (`json` and `serve`) along with `helmet` stand out to me as possibilities.

My argument against pulling *all* of them into core, even in the future, is the sheer number of dependencies. There are about 16 dependencies for databases alone, plus another 15 or so. We have already had some issues with dependency versions where it might be nice to untether the release cycle of `rocket_contrib` from that of `rocket`, for example allowing `rocket_contrib` to advance to `2.0` and `3.0` while `rocket` remains at `1.0`. It may even make sense to further break out integrations with a short major version cycle (`rusqlite` is a example I can think of right now).

One data point I am very interested in is how many people are using which features. This is difficult to gauge because many projects using Rocket probably aren't public, but we might be able to get stats on public repositories at the very least.

**Comment by jhpratt** [comment]
Yeah, this definitely wouldn't be done for 0.5; that wasn't my intent. It's more of a "hey, here's an idea I had". I'd tentatively put the "minimal" migration (consisting of `json`, `serve` and `helmet`) at 0.6.

The point you bring up about versioning is certainly an interesting one. Features whose dependences are stable or Rocket-controlled would be the primary candidates for migration, as changes would either be minimal (for stables) or controllable (for Rocket-controlled).

Excluding `futures-preview` for obvious reasons, `json`, `serve` and `helmet` would all fall under this classification.
- `json` depends on stable `serde` and `serde_json`
- `serve` is entirely controlled by Rocket, having no dependencies
- `helmet` depends on the unmaintained `time` crate, though this would be trivially removed at a slight cost to readability (`time::Duration::days(30)` would become `std::time::Duration::from_secs(30 * 86_400)`, for example). It is otherwise entirely controlled by Rocket.

Templates (both handlebars and Tera<sup>β</sup>) are stable, which is the only other feature that falls under this pseudo-criteria.

<sup>For the purposes of this comment, when I refer to stable I mean 1.0 and higher</sup>

**Comment by TheBestJohn** [comment]
Weighing in from my perspective, and by that I mean take it with a grain of salt. I see rocket filling a pseudo-kitchen-sink magically-abstract-away repetitive bootstrapping sort of niche. I understand and appreciate the `rocket-contrib` viewpoint of "vetted but not core team developed" but that distinction doesn't necessarily need to be as physical in nature as a separate crate. 

Objectively, can `rocket_contrib` be useful to any application other than rocket? Is it a standalone library or an optional component of Rocket? I do see the paradigm of many popular crates breaking up functionality into separate crates (tokio is the first that comes to mind) but the question I guess is is that even a "good" way to do things?

**Comment by jebrosen** [comment]
> Objectively, can rocket_contrib be useful to any application other than rocket?

Not really, but that shouldn't be the deciding factor.

---

(Disclaimer: I don't know Sergio's original intentions for `rocket_contrib` and I may disagree with those in some respects. I also describe an ideal situation of sorts; reality sometimes misses the mark.)

When I refer to `rocket`, I usually mean everything in `core/`: `rocket`, `rocket_http`, `rocket_codegen`.

IMO `rocket` should always be solid and well audited, with only high-quality dependencies. Trusting `rocket` should be an easy choice to make: it should be API stable, it should be reliable at runtime (predictable, resilient, reasonable resource usage, etc.) It should not have unnecessary dependencies. Auditing `rocket` for correctness should not require checking the usage of (for example) a database crate, as it does not itself do anything interesting with databases. Provided request guards and responders should be more or less universally applicable.

`rocket_contrib`, on the other hand, has more wiggle room. It depends on databases, template engines, filesystem watchers, (de)serialization libraries, and more. The release cycles and stability policies of those crates may not match with rocket's; they may not be as thoroughly tested on all platforms, etc. In the future, I do think the release cycle of `rocket_contrib` should be faster than that of `rocket`. I've even toyed with the idea of breaking database support out into a separate crate, just because of the dependency update cycles.

In all, I think the separation is important. We might merge or split functionality differently, but I think there will always be some features that are of high enough quality and usefulness to officially support, but not core enough to pull into every single application that uses `rocket`.

**Comment by TheBestJohn** [comment]
You've swayed me. That totally makes sense to me now. It's a matter of what should be considered "core". Some arguments to be made about what belongs there "space helmet would seem to fit the bill" but I get why something like Templates or diesel would not be a good fit. It also makes sense that say, in a few years, nobody is using diesel anymore because there's some fundamental flaw. It would mean that Rocket Core would have to excise that from itself. Messy.

**Comment by TheBestJohn** [comment]
Coming back to this now, it makes even more sense. With Async on the horizon for 0.5.0, it looks like Deadpool is going to fill a part of the Database hole for a lot of users. Moving Diesel integration to core would have have complicated this a bit.

**Comment by jhpratt** [comment]
Update on this: I no longer believe everything should be merged in. However, the aforementioned list (json, serve, and helmet) I still believe could and should be merged into core at some point.

**Comment by Aaditya-Sahay** [comment]
I speak from the perspective of the user of rocket, and @jhpratt 's point make a lot of sense to me. I would expect certain feature present right now in rocket_contrib to be a part of the core itself, (and certainly did expect that to be the case when I initially used the rocket framework) json for example, as has been mentioned above seems like a great candidate. At the same time, I find myself totally comfortable with the present structure and organisation of the crates. 
Hope this is of some use :)
